gateway-service-knoxtoken/src/main/java/org/apache/knox/gateway/service/knoxtoken/TokenResource.java

```diff
@@ -720,19 +705,21 @@ private Response getAuthenticationToken() {
     String createdBy = null;
     // checking the doAs user only makes sense if tokens are managed (this is where we store the userName information)
     // and if impersonation is enabled
-    if (impersonationEnabled && tokenStateService != null) {
-      final String doAsUser = request.getParameter(QUERY_PARAMETER_DOAS);
-      if (doAsUser != null && !doAsUser.equals(userName)) {
-        try {
-          //this call will authorize the doAs request
-          AuthFilterUtils.authorizeImpersonationRequest(request, doAsUser, getTopologyName(), TokenServiceDeploymentContributor.ROLE);
-          createdBy = userName;
-          userName = doAsUser;
-          log.tokenImpersonationSuccess(createdBy, doAsUser);
-        } catch (AuthorizationException e) {
-          log.tokenImpersonationFailed(e);
-          return Response.status(Response.Status.FORBIDDEN).entity("{ \"" + e.getMessage() + "\" }").build();
+    if (tokenStateService != null) {
+      final String realUserName = (String) request.getAttribute(AuthFilterUtils.REAL_USER_NAME_ATTRIBUTE);
+      final Subject subject = SubjectUtils.getCurrentSubject();
+      if (subject != null && SubjectUtils.isImpersonating(subject)) {
+        String primaryPrincipalName = SubjectUtils.getPrimaryPrincipalName(subject);
+        String impersonatedPrincipalName = SubjectUtils.getImpersonatedPrincipalName(subject);
+        if (!primaryPrincipalName.equals(impersonatedPrincipalName)) {
+          createdBy = primaryPrincipalName;
+          userName = impersonatedPrincipalName;
+          log.tokenImpersonationSuccess(createdBy, userName);
         }
+      } else if (StringUtils.isNotBlank(realUserName) && !realUserName.equals(userName)) {
+        // real user name is set by HadoopAuth filter for impersonated requests (part of 'doAs' processing)
+        createdBy = realUserName;
```

Comment:
I agree that using the Subject Principals as the vehicle for communicating this information would be ideal.

Reply:
@pzampino , @lmccay - I agree too. However, the way how the HadoopAuthFilter and the IdentityAssertionFilter handle impersonation is different:
- the first uses a wrapped HTTP request down the line, where the `getRemoteUser` and `getUserPrincipal` methods are overridden
- whereas the second uses `Subject.doAs` with the configured principals

I also prefer the 2nd way but given we are close to the release I'm not sure if changing the `HadoopAuthFilter` behavior to use `Subject.doAs` is a good idea.

Reply:
Subject.doAs should be used everywhere. Let me check the HadoopAuthProvider code. This is a fundamental aspect of the Knox provider and service separation though. By the time the request gets to a service the security context (Subject) needs to have everything that is needed. All provider specific details are normalized into Knox standard Principals. If the Knox standard needs to change to accommodate things then it needs to be done in there. I only point this out because I don't think this has been clear enough which I think is how we got here.

Reply:
https://github.com/apache/knox/blob/master/gateway-provider-security-hadoopauth/src/main/java/org/apache/knox/gateway/hadoopauth/filter/HadoopAuthPostFilter.java#L120